specreduce/extract.py

```diff
@@ -143,3 +146,174 @@ def _ap_weight_image(trace, width, disp_axis, crossdisp_axis, image_shape):
+            unmasked. [default: None]
+
+        unit : `~astropy.units.core.Unit` or str, optional
+            (Only used if `image` is not an NDData object.)
+            The associated unit for the data in `image`. If blank,
+            fluxes are interpreted as unitless. [default: None]
+
+
+        Returns
+        -------
+        spec_1d : `~specutils.Spectrum1D`
+            The final, Horne extracted 1D spectrum.
+        """
+        # handle image and associated data based on image's type
+        if isinstance(image, NDData):
+            img = np.ma.array(image.data, mask=image.mask)
+            unit = image.unit if image.unit is not None else u.Unit()
+
+            if image.uncertainty is not None:
+                # prioritize NDData's uncertainty over variance argument
+                if image.uncertainty.uncertainty_type == 'var':
+                    variance = image.uncertainty.array
+                elif image.uncertainty.uncertainty_type == 'std':
+                    # NOTE: CCDData defaults uncertainties given as pure arrays
+                    # to std and logs a warning saying so upon object creation.
+                    # should we remind users again here?
+                    warnings.warn("image NDData object's uncertainty "
+                                  "interpreted as standard deviation. if "
+                                  "incorrect, use VarianceUncertainty when "
+                                  "assigning image object's uncertainty.")
+                    variance = image.uncertainty.array**2
+                elif image.uncertainty.uncertainty_type == 'ivar':
+                    variance = 1 / image.uncertainty.array
+                else:
+                    # other options are InverseVariance and UnknownVariance
+                    raise ValueError("image NDData object has unexpected "
+                                     "uncertainty type. instead, try "
+                                     "VarianceUncertainty or StdDevUncertainty.")
+            else:
+                # ignore variance arg to focus on updating NDData object
+                raise ValueError('image NDData object lacks uncertainty')
+
+        else:
+            if any(arg is None for arg in (variance, mask, unit)):
+                raise ValueError('if image is a numpy array, the variance, '
+                                 'mask, and unit arguments must be specified. '
+                                 'consider wrapping that information into one '
+                                 'object by instead passing an NDData image.')
+            if image.shape != variance.shape:
+                raise ValueError('image and variance shapes must match')
+            if image.shape != mask.shape:
+                raise ValueError('image and mask shapes must match')
+
+            # fill in non-required arguments if empty
+            if mask is None:
+                mask = np.ma.masked_invalid(image)
```

Comment:
@ojustino - this came up today when trying to use Horne extraction.  The first if statement raises an error if mask is not passed, but later there is a check to fallback if mask is None.  Can both mask and unit (default to pixel) be optional when passing image as an array instead of an NDData object?  (If so, maybe this should be a follow-up issue since this PR has since been merged, I just wanted to attach it to the lines of code somewhere to have a breadcrumb)

Reply:
I believe we settled on `mask` and `unit` being optional for array-type images, so the `if any()` should be updated to reflect that. I can open a pull request for it you'd like to copy over some of what you've written in a new issue.

I agree that `mask` check should be moved earlier.